File: android/app/src/main/kotlin/app/familyagent/android/data/SettingsStore.kt
package app.familyagent.android.data

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

private val Context.dataStore by preferencesDataStore(name = "family_agent_settings")
private val SERVER_URL_KEY = stringPreferencesKey("server_url")
private val AUTH_TOKEN_KEY = stringPreferencesKey("auth_token")
private val USER_NAME_KEY = stringPreferencesKey("user_display_name")
private val SERVER_NAME_KEY = stringPreferencesKey("server_name")
private val TASK_VIEW_KEY = stringPreferencesKey("task_view")
private val AUTO_READ_KEY = booleanPreferencesKey("auto_read_replies")
private val MIC_ON_LEFT_KEY = booleanPreferencesKey("mic_button_on_left")
// "Remember me" on the login screen — keyed by server URL (below the plain
// "current session" keys above) so switching servers doesn't leak one home's
// saved login into another's fields. Plaintext at rest, same as AUTH_TOKEN_KEY
// above: this DataStore is already the app's private sandboxed storage and
// isn't field-level encrypted for the token either, so this isn't a new
// exposure — just app-sandbox protected, like everything else here.
private val REMEMBERED_LOGIN_URL_KEY = stringPreferencesKey("remembered_login_url")
private val REMEMBERED_USERNAME_KEY = stringPreferencesKey("remembered_username")
private val REMEMBERED_PASSWORD_KEY = stringPreferencesKey("remembered_password")
private val TASK_VIEWS = listOf("list", "day", "3day", "week", "month")

// Prefill for the manual-address field only. The normal path is LAN discovery
// (ServerDiscovery) — this app talks only to a server the user picked or typed.
const val DEFAULT_SERVER_URL = "http://192.168.1.2:4173"

/** What we remember about the signed-in session between launches. */
data class Session(
    val serverUrl: String,
    val token: String,
    val displayName: String,
    val serverName: String,
)

class SettingsStore(private val context: Context) {
    val serverUrl = context.dataStore.data.map { it[SERVER_URL_KEY] ?: DEFAULT_SERVER_URL }

    val session = context.dataStore.data.map { prefs ->
        val url = prefs[SERVER_URL_KEY]
        val token = prefs[AUTH_TOKEN_KEY]
        if (url != null && token != null) {
            Session(url, token, prefs[USER_NAME_KEY].orEmpty(), prefs[SERVER_NAME_KEY].orEmpty())
        } else null
    }

    suspend fun setServerUrl(url: String) {
        context.dataStore.edit { it[SERVER_URL_KEY] = url.trimEnd('/') }
    }

    /** Which Tasks view to show: list | day | 3day | week | month. Survives sign-out. */
    val taskView = context.dataStore.data.map {
        when (val v = it[TASK_VIEW_KEY]) {
            null -> "week"          // new default
            "calendar" -> "month"  // legacy value from the first version
            in TASK_VIEWS -> v!!
            else -> "week"
        }
    }

    suspend fun setTaskView(view: String) {
        context.dataStore.edit { it[TASK_VIEW_KEY] = view }
    }

    /** Read new assistant replies aloud automatically. Survives sign-out. */
    val autoRead = context.dataStore.data.map { it[AUTO_READ_KEY] ?: false }

    suspend fun setAutoRead(on: Boolean) {
        context.dataStore.edit { it[AUTO_READ_KEY] = on }
    }

    /**
     * Which side of the composer the hold-to-talk mic sits on: true = left of
     * the text field (left-handed reach), false = right, next to Send (default).
     * Device-local, survives sign-out.
     */
    val micOnLeft = context.dataStore.data.map { it[MIC_ON_LEFT_KEY] ?: false }

    suspend fun setMicOnLeft(on: Boolean) {
        context.dataStore.edit { it[MIC_ON_LEFT_KEY] = on }
    }

    suspend fun saveSession(serverUrl: String, token: String, displayName: String, serverName: String) {
        context.dataStore.edit {
            it[SERVER_URL_KEY] = serverUrl.trimEnd('/')
            it[AUTH_TOKEN_KEY] = token
            it[USER_NAME_KEY] = displayName
            it[SERVER_NAME_KEY] = serverName
        }
    }

    suspend fun clearSession() {
        context.dataStore.edit {
            it.remove(AUTH_TOKEN_KEY)
            it.remove(USER_NAME_KEY)
        }
    }

    /** A saved username/password for [serverUrl], if "Remember me" was checked
     * on a previous sign-in there — the login screen prefills from it. */
    suspend fun rememberedLogin(serverUrl: String): Pair<String, String>? {
        val prefs = context.dataStore.data.first()
        if (prefs[REMEMBERED_LOGIN_URL_KEY] != serverUrl.trimEnd('/')) return null
        val u = prefs[REMEMBERED_USERNAME_KEY] ?: return null
        val p = prefs[REMEMBERED_PASSWORD_KEY] ?: return null
        return u to p
    }

    suspend fun saveRememberedLogin(serverUrl: String, username: String, password: String) {
        context.dataStore.edit {
            it[REMEMBERED_LOGIN_URL_KEY] = serverUrl.trimEnd('/')
            it[REMEMBERED_USERNAME_KEY] = username
            it[REMEMBERED_PASSWORD_KEY] = password
        }
    }

    suspend fun clearRememberedLogin() {
        context.dataStore.edit {
            it.remove(REMEMBERED_LOGIN_URL_KEY)
            it.remove(REMEMBERED_USERNAME_KEY)
            it.remove(REMEMBERED_PASSWORD_KEY)
        }
    }
}

File: ios/FamilyAgent/Networking/SettingsStore.swift
import Foundation

/// Persisted session + client-local preferences — the iOS mirror of
/// `android/.../data/SettingsStore.kt`. Server URL + names in `UserDefaults`,
/// the bearer token in the Keychain.
struct SettingsStore {
    private let defaults = UserDefaults.standard
    private static let tokenAccount = "session"

    private enum K {
        static let serverURL = "server_url"
        static let serverName = "server_name"
        static let userName = "user_display_name"
        static let taskView = "task_view"
        static let autoRead = "auto_read_replies"
        static let micOnLeft = "mic_button_on_left"
        static let recentServers = "recent_servers"
    }

    struct Session {
        var serverURL: String
        var token: String
        var serverName: String
        var userName: String
    }

    /// A server this device has successfully connected to before — shown on
    /// the discovery screen as a one-tap reconnect. This is what makes a
    /// Tailscale (or any off-LAN) address usable at all after the first time:
    /// Bonjour/mDNS can't find it again on its own (multicast doesn't cross
    /// a tailnet), so remembering it here is the fix, not a better scan.
    struct RecentServer: Codable, Identifiable, Equatable {
        var name: String
        var url: String
        var id: String { url }
    }

    /// Most-recently-used first, deduped by URL, capped at 5.
    var recentServers: [RecentServer] {
        get {
            guard let data = defaults.data(forKey: K.recentServers),
                  let list = try? JSONDecoder().decode([RecentServer].self, from: data) else { return [] }
            return list
        }
        nonmutating set {
            guard let data = try? JSONEncoder().encode(Array(newValue.prefix(5))) else { return }
            defaults.set(data, forKey: K.recentServers)
        }
    }

    func addRecentServer(name: String, url: String) {
        var list = recentServers
        list.removeAll { $0.url == url }
        list.insert(RecentServer(name: name, url: url), at: 0)
        recentServers = list
    }

    var session: Session? {
        guard let url = defaults.string(forKey: K.serverURL),
              let token = Keychain.get(account: Self.tokenAccount) else { return nil }
        return Session(
            serverURL: url,
            token: token,
            serverName: defaults.string(forKey: K.serverName) ?? "",
            userName: defaults.string(forKey: K.userName) ?? ""
        )
    }

    func saveSession(serverURL: String, token: String, serverName: String, userName: String) {
        defaults.set(serverURL, forKey: K.serverURL)
        defaults.set(serverName, forKey: K.serverName)
        defaults.set(userName, forKey: K.userName)
        Keychain.set(token, account: Self.tokenAccount)
    }

    func clearSession() {
        Keychain.delete(account: Self.tokenAccount)
        defaults.removeObject(forKey: K.userName)
    }

    func setServerURL(_ url: String) { defaults.set(url, forKey: K.serverURL) }
    var serverURL: String { defaults.string(forKey: K.serverURL) ?? "" }

    var taskView: String {
        get {
            let v = defaults.string(forKey: K.taskView) ?? "week"
            return v == "calendar" ? "month" : v
        }
        nonmutating set { defaults.set(newValue, forKey: K.taskView) }
    }

    var autoRead: Bool {
        get { defaults.bool(forKey: K.autoRead) }
        nonmutating set { defaults.set(newValue, forKey: K.autoRead) }
    }

    /// Which side of the composer the hold-to-talk mic sits on. Default `false`
    /// = right (next to Send); `true` puts it left of the text field for
    /// left-handed reach. Device-local, like `autoRead`.
    var micOnLeft: Bool {
        get { defaults.bool(forKey: K.micOnLeft) }
        nonmutating set { defaults.set(newValue, forKey: K.micOnLeft) }
    }
}
